single_kernel_mongo/events/cluster.py

```diff
@@ -0,0 +1,156 @@
+    DeferrableError,
+    DeferrableFailedHookChecksError,
+    NonDeferrableFailedHookChecksError,
+    WaitingForSecretsError,
+)
+from single_kernel_mongo.lib.charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseProviderEventHandlers,
+    DatabaseRequestedEvent,
+    DatabaseRequirerEventHandlers,
+)
+from single_kernel_mongo.utils.event_helpers import defer_event_with_info_log
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.managers.mongodb_operator import MongoDBOperator
+    from single_kernel_mongo.managers.mongos_operator import MongosOperator
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterConfigServerEventHandler(Object):
+    """Event Handler for managing config server side events."""
+
+    def __init__(self, dependent: MongoDBOperator):
+        self.dependent = dependent
+        self.charm = self.dependent.charm
+        self.manager = self.dependent.cluster_manager
+        self.relation_name = self.manager.relation_name
+        super().__init__(parent=self.manager, key=dependent.cluster_manager.relation_name)
+
+        self.database_provider_events = DatabaseProviderEventHandlers(
+            self.charm, self.manager.data_interface
+        )
+
+        self.framework.observe(
+            self.database_provider_events.on.database_requested, self._on_database_requested
+        )
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_changed, self._on_relation_event
+        )
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_departed,
+            self.dependent.check_relation_broken_or_scale_down,
+        )
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_broken, self._on_relation_broken_event
+        )
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent):
+        """Relation joined events."""
+        try:
+            self.manager.share_secret_to_mongos(event.relation)
+        except DeferrableFailedHookChecksError as e:
+            logger.info("Skipping database requested event: hook checks did not pass.")
+            defer_event_with_info_log(logger, event, str(type(event)), str(e))
+        except NonDeferrableFailedHookChecksError as e:
+            logger.info(f"Skipping {str(type(event))}: {str(e)}")
+
+    def _on_relation_event(self, event: RelationChangedEvent):
+        """Handle relation changed and relation broken events."""
+        try:
+            self.manager.update_keyfile_and_hosts_on_mongos(event.relation)
+        except DeferrableFailedHookChecksError as e:
+            defer_event_with_info_log(logger, event, str(type(event)), str(e))
+        except NonDeferrableFailedHookChecksError as e:
+            logger.info(f"Skipping {str(type(event))}: {str(e)}")
+
+    def _on_relation_broken_event(self, event: RelationBrokenEvent):
+        try:
+            self.manager.on_relation_broken(event.relation)
```

Comment:
`relation_broken` doesn't seem like a method name a manager should have, could you be more explicit / business logic oriented here?

Reply:
Sure, let me rename this!